File: sdk/android/src/main/java/com/hotpatch/OTAUpdateManager.kt
package com.hotpatch

import android.content.Context
import android.util.Log
import java.io.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import okhttp3.*
import org.json.JSONObject

object OTAUpdateManager {
    private const val TAG = "HotPatch"
    private var apiUrl: String = ""
    private var appId: String = ""
    private var channel: String = "production"
    private var encryptionKey: String? = null
    private var signingPublicKey: String? = null
    private var pinnedDomain: String? = null
    private var pinnedHash: String? = null

    private var client = OkHttpClient()
    private val executor: ExecutorService = Executors.newSingleThreadExecutor()

    fun setup(
            url: String,
            id: String,
            ch: String,
            key: String? = null,
            signingKey: String? = null
    ) {
        apiUrl = url
        appId = id
        channel = ch
        encryptionKey = key
        signingPublicKey = signingKey
        rebuildClient()
    }

    fun setupCertificatePinning(domain: String, hash: String) {
        pinnedDomain = domain
        pinnedHash = hash
        rebuildClient()
    }

    private fun rebuildClient() {
        val builder = OkHttpClient.Builder()
        if (pinnedDomain != null && pinnedHash != null) {
            val pinner =
                    okhttp3.CertificatePinner.Builder().add(pinnedDomain!!, pinnedHash!!).build()
            builder.certificatePinner(pinner)
        }
        client = builder.build()
    }

    fun getBundlePath(context: Context): String? {
        val otaDir = File(context.filesDir, "ota")
        if (!otaDir.exists()) otaDir.mkdirs()

        // Increment crash count on every launch
        val crashCountFile = File(otaDir, "crash_count.json")
        var count = 0
        if (crashCountFile.exists()) {
            try {
                count = JSONObject(crashCountFile.readText()).optInt("count", 0)
            } catch (e: Exception) {}
        }

        if (count >= 2) {
            Log.e(TAG, "Boot loop detected! Rolling back to last stable version.")
            rollback(context)
            return null
        }

        // Write incremented count
        val json = JSONObject()
        json.put("count", count + 1)
        crashCountFile.writeText(json.toString())

        val otaBundle = File(context.filesDir, "ota/bundle/index.android.bundle")
        return if (otaBundle.exists() && isVerified(context)) {
            otaBundle.absolutePath
        } else {
            null
        }
    }

    private fun isVerified(context: Context): Boolean {
        val metadataFile = File(context.filesDir, "ota/metadata.json")
        if (!metadataFile.exists()) return false
        try {
            val json = JSONObject(metadataFile.readText())
            return json.has("version") && json.has("hash")
        } catch (e: Exception) {
            return false
        }
    }

    fun markSuccess(context: Context) {
        val crashCountFile = File(context.filesDir, "ota/crash_count.json")
        if (crashCountFile.exists()) {
            try {
                val json = JSONObject()
                json.put("count", 0)
                crashCountFile.writeText(json.toString())
            } catch (e: Exception) {}
        }
    }

    private fun rollback(context: Context) {
        val bundleDir = File(context.filesDir, "ota/bundle")
        val previousDir = File(context.filesDir, "ota/previous")

        // Load metadata to get releaseId before rollback
        var releaseId = ""
        val metadataFile = File(context.filesDir, "ota/metadata.json")
        if (metadataFile.exists()) {
            try {
                releaseId = JSONObject(metadataFile.readText()).optString("release_id")
            } catch (e: Exception) {}
        }

        if (previousDir.exists()) {
            if (bundleDir.exists()) bundleDir.deleteRecursively()
            previousDir.renameTo(bundleDir)

            // Clean up metadata or update it to previous if available
            // For now, just mark rollback in backend
            if (releaseId.isNotEmpty()) {
                reportInstallation(context, releaseId, "rolled_back", false, 0)
            }
        } else {
            bundleDir.deleteRecursively()
            metadataFile.delete()
        }
    }

    fun checkForUpdate(
            context: Context,
            currentVersion: String,
            deviceId: String,
            callback: (JSONObject?) -> Unit
    ) {
        val prefs = context.getSharedPreferences("HotPatch", Context.MODE_PRIVATE)
        prefs.edit().putString("deviceId", deviceId).apply()

        executor.execute {
            try {
                val url =
                        "$apiUrl/update/check?appId=$appId&deviceId=$deviceId&version=$currentVersion&platform=android&channel=$channel"
                val request = Request.Builder().url(url).header("X-App-Key", appId).build()
                val response = client.newCall(request).execute()
                val body = response.body?.string()

                if (body != null) {
                    val json = JSONObject(body)
                    if (json.optBoolean("updateAvailable")) {
                        callback(json)
                    } else {
                        callback(null)
                    }
                } else {
                    callback(null)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Update check failed", e)
                callback(null)
            }
        }
    }

    fun downloadAndApply(context: Context, updateJson: JSONObject) {
        executor.execute {
            try {
                val bundleUrl = updateJson.getString("bundleUrl")

                // Prepare dirs
                val bundleDir = File(context.filesDir, "ota/bundle")
                val previousDir = File(context.filesDir, "ota/previous")
                val pendingDir = File(context.filesDir, "ota/pending")
                if (!pendingDir.exists()) pendingDir.mkdirs()

                // Download
                val downloadedFile = File(pendingDir, "bundle.tmp")
                val request = Request.Builder().url(bundleUrl).build()
                val response = client.newCall(request).execute()

                val fos = FileOutputStream(downloadedFile)
                fos.write(response.body?.bytes())
                fos.close()

                var finalZip = downloadedFile

                // Step A: Handle Differential Patch
                if (updateJson.optBoolean("isPatch")) {
                    Log.i(TAG, "Applying differential patch...")
                    val currentBundleZip = File(context.filesDir, "ota/bundle.zip")
                    if (!currentBundleZip.exists()) {
                        Log.e(TAG, "Base bundle missing for patch!")
                        return@execute
                    }
                    val patchedZip = File(pendingDir, "bundle.patched.zip")
                    applyPatch(currentBundleZip, downloadedFile, patchedZip)
                    finalZip = patchedZip
                }

                // Step C: Verify Integrity
                val expectedHash = updateJson.optString("hash")
                if (expectedHash.isNotEmpty() && !HashUtils.verifyHash(finalZip, expectedHash)) {
                    Log.e(TAG, "Hash verification failed!")
                    return@execute
                }

                val signature = updateJson.optString("signature")
                if (signature.isNotEmpty() && signingPublicKey != null) {
                    if (!SignatureVerifier.verifySignature(finalZip, signature, signingPublicKey!!)
                    ) {
                        Log.e(TAG, "Signature verification failed!")
                        return@execute
                    }
                    Log.i(TAG, "Authenticity verified ✓")
                }

                // Step B: Handle Decryption (after verification, before extraction)
                if (updateJson.optBoolean("isEncrypted")) {
                    Log.i(TAG, "Decrypting bundle...")
                    val decryptedZip = File(pendingDir, "bundle.decrypted.zip")
                    decryptFile(finalZip, decryptedZip, encryptionKey)
                    finalZip = decryptedZip
                }

                // Move current to previous
                if (bundleDir.exists()) {
                    if (previousDir.exists()) previousDir.deleteRecursively()
                    bundleDir.renameTo(previousDir)
                }

                // Extract ZIP
                bundleDir.mkdirs()
                unzip(finalZip, bundleDir)

                // Save metadata
                val metadata =
                        JSONObject().apply {
                            put("version", updateJson.getString("version"))
                            put("hash", updateJson.getString("hash"))
                            put(
                                    "release_id",
                                    updateJson.optString("id", updateJson.optString("release_id"))
                            )
                            put("installed_at", System.currentTimeMillis())
                        }
                File(context.filesDir, "ota/metadata.json").writeText(metadata.toString())

                // Keep the final zip for future differential updates
                finalZip.copyTo(File(context.filesDir, "ota/bundle.zip"), overwrite = true)

                Log.i(TAG, "Update Applied Successfully.")

                // Report installation
                val releaseId = updateJson.optString("id", updateJson.optString("release_id"))
                val isPatch = updateJson.optBoolean("isPatch")
                val downloadSize = downloadedFile.length()

                reportInstallation(context, releaseId, "applied", isPatch, downloadSize)

                if (updateJson.optBoolean("mandatory")) {
                    Log.i(TAG, "Mandatory update! Triggering reload...")
                    triggerReload(context)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Download failed", e)
            }
        }
    }

    private fun reportInstallation(
            context: Context,
            releaseId: String,
            status: String,
            isPatch: Boolean,
            downloadSize: Long
    ) {
        val prefs = context.getSharedPreferences("HotPatch", Context.MODE_PRIVATE)
        val deviceId = prefs.getString("deviceId", null) ?: return

        val url = "$apiUrl/installations"
        val body =
                JSONObject().apply {
                    put("device_id", deviceId)
                    put("release_id", releaseId)
                    put("status", status)
                    put("is_patch", isPatch)
                    put("download_size", downloadSize)
                }

        val request =
                Request.Builder()
                        .url(url)
                        .post(
                                RequestBody.create(
                                        MediaType.parse("application/json"),
                                        body.toString()
                                )
                        )
                        .addHeader("X-App-Key", appId)
                        .build()

        client.newCall(request)
                .enqueue(
                        object : Callback {
                            override fun onFailure(call: Call, e: IOException) {
                                Log.e(TAG, "Reporting failed", e)
                            }
                            override fun onResponse(call: Call, response: Response) {
                                response.close()
                            }
                        }
                )
    }

    private fun applyPatch(oldFile: File, patchFile: File, outputFile: File) {
        Log.i(TAG, "Applying binary patch from ${oldFile.name}")
        PatchUtils.applyPatch(oldFile, patchFile, outputFile)
    }

    private fun decryptFile(inputFile: File, outputFile: File, keyHex: String?) {
        if (keyHex == null) throw Exception("Decryption key missing")

        val data = inputFile.readBytes()
        val nonce = data.sliceArray(0 until 12)
        val ciphertext = data.sliceArray(12 until data.size)

        val key = hexToBytes(keyHex)
        val secretKey = javax.crypto.spec.SecretKeySpec(key, "AES")
        val cipher = javax.crypto.Cipher.getInstance("AES/GCM/NoPadding")
        val gcmSpec = javax.crypto.spec.GCMParameterSpec(128, nonce)

        cipher.init(javax.crypto.Cipher.DECRYPT_MODE, secretKey, gcmSpec)
        val decrypted = cipher.doFinal(ciphertext)

        outputFile.writeBytes(decrypted)
    }

    private fun unzip(zipFile: File, targetDir: File) {
        val zipIn = java.util.zip.ZipInputStream(java.io.FileInputStream(zipFile))
        var entry = zipIn.nextEntry
        while (entry != null) {
            val filePath = File(targetDir, entry.name)
            if (!entry.isDirectory) {
                val fos = java.io.FileOutputStream(filePath)
                zipIn.copyTo(fos)
                fos.close()
            } else {
                filePath.mkdirs()
            }
            zipIn.closeEntry()
            entry = zipIn.nextEntry
        }
        zipIn.close()
    }

    private fun hexToBytes(hex: String): ByteArray {
        val result = ByteArray(hex.length / 2)
        for (i in 0 until hex.length step 2) {
            result[i / 2] = hex.substring(i, i + 2).toInt(16).toByte()
        }
        return result
    }

    private fun triggerReload(context: Context) {
        try {
            val activity =
                    (context as? android.app.Activity)
                            ?: (context as? android.content.ContextWrapper)?.baseContext as?
                                    android.app.Activity

            activity?.runOnUiThread {
                // This is a simplified way to reload.
                // A better way is using ReactInstanceManager.
                activity.recreate()
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to reload activity", e)
        }
    }
}

File: sdk/ios/OTAUpdateManager.swift
import Foundation
#if canImport(CryptoKit)
import CryptoKit
#endif

@objc(OTAUpdateManager)
public class OTAUpdateManager: NSObject, URLSessionDelegate {
    private var apiUrl: String = ""
    private var appId: String = ""
    private var channel: String = "production"
    private var encryptionKey: String? = nil
    private var signingPublicKey: String? = nil
    private var pinnedDomain: String? = nil
    private var pinnedPublicKeyHash: String? = nil
    private let TAG = "HotPatch"
    
    private var session: URLSession!
    
    @objc public static let shared = OTAUpdateManager()
    
    @objc public func setup(url: String, id: String, ch: String, key: String? = nil, signingKey: String? = nil) {
        self.apiUrl = url
        self.appId = id
        self.channel = ch
        self.encryptionKey = key
        self.signingPublicKey = signingKey
        rebuildSession()
    }
    
    @objc public func setupCertificatePinning(domain: String, hash: String) {
        self.pinnedDomain = domain
        self.pinnedPublicKeyHash = hash
        rebuildSession()
    }
    
    private func rebuildSession() {
        let config = URLSessionConfiguration.default
        self.session = URLSession(configuration: config, delegate: self, delegateQueue: nil)
    }
    
    @objc public func getBundleURL() -> URL? {
        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
        let otaDir = paths[0].appendingPathComponent("ota")
        let crashFile = otaDir.appendingPathComponent("crash_count.json")
        
        if !FileManager.default.fileExists(atPath: otaDir.path) {
            try? FileManager.default.createDirectory(at: otaDir, withIntermediateDirectories: true)
        }
        
        var count = 0
        if let data = try? Data(contentsOf: crashFile),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Int] {
            count = json["count"] ?? 0
        }
        
        if count >= 2 {
            print("[\(TAG)] Boot loop detected! Rolling back.")
            rollback()
            return nil
        }
        
        // Save incremented count
        let newJson = ["count": count + 1]
        if let newData = try? JSONSerialization.data(withJSONObject: newJson) {
            try? newData.write(to: crashFile)
        }
        
        let bundleURL = otaDir.appendingPathComponent("bundle/main.jsbundle")
        if FileManager.default.fileExists(atPath: bundleURL.path) && isVerified() {
            return bundleURL
        }
        return nil
    }
    
    private func isVerified() -> Bool {
        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
        let metadataFile = paths[0].appendingPathComponent("ota/metadata.json")
        guard let data = try? Data(contentsOf: metadataFile),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return false
        }
        return json["version"] != nil && json["hash"] != nil
    }
    
    @objc public func markSuccess() {
        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
        let crashFile = paths[0].appendingPathComponent("ota/crash_count.json")
        let json = ["count": 0]
        if let data = try? JSONSerialization.data(withJSONObject: json) {
            try? data.write(to: crashFile)
        }
    }
    
    private func rollback() {
        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
        let otaDir = paths[0].appendingPathComponent("ota")
        let bundleDir = otaDir.appendingPathComponent("bundle")
        let previousDir = otaDir.appendingPathComponent("previous")
        let metadataFile = otaDir.appendingPathComponent("metadata.json")
        
        var releaseId = ""
        if let data = try? Data(contentsOf: metadataFile),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            releaseId = json["release_id"] as? String ?? ""
        }

        if FileManager.default.fileExists(atPath: previousDir.path) {
            try? FileManager.default.removeItem(at: bundleDir)
            try? FileManager.default.moveItem(at: previousDir, to: bundleDir)
            
            if !releaseId.isEmpty {
                reportInstallation(releaseId: releaseId, status: "rolled_back", isPatch: false, downloadSize: 0)
            }
        } else {
            try? FileManager.default.removeItem(at: bundleDir)
            try? FileManager.default.removeItem(at: metadataFile)
        }
    }
    
    func checkForUpdate(version: String, deviceId: String, completion: @escaping ([String: Any]?) -> Void) {
        UserDefaults.standard.set(deviceId, forKey: "HotPatch_DeviceId")
        let urlString = "\(apiUrl)/update/check?appId=\(appId)&deviceId=\(deviceId)&version=\(version)&platform=ios&channel=\(channel)"
        guard let url = URL(string: urlString) else {
            completion(nil)
            return
        }
        
        var request = URLRequest(url: url)
        request.addValue(appId, forHTTPHeaderField: "X-App-Key")
        
        session.dataTask(with: request) { data, response, error in
            guard let data = data, error == nil else {
                completion(nil)
                return
            }
            
            if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let available = json["updateAvailable"] as? Bool, available {
                completion(json)
            } else {
                completion(nil)
            }
        }.resume()
    }
    
    func downloadAndApply(updateJson: [String: Any], completion: @escaping (Bool) -> Void) {
        let bundleUrlStr = updateJson["bundleUrl"] as? String ?? ""
        guard let bundleUrl = URL(string: bundleUrlStr) else {
            completion(false)
            return
        }
        
        session.downloadTask(with: bundleUrl) { location, response, error in
            guard let location = location, error == nil else {
                completion(false)
                return
            }
            
            do {
                let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
                let otaDir = paths[0].appendingPathComponent("ota")
                let pendingDir = otaDir.appendingPathComponent("pending")
                let bundleDir = otaDir.appendingPathComponent("bundle")
                let previousDir = otaDir.appendingPathComponent("previous")
                
                if !FileManager.default.fileExists(atPath: pendingDir.path) {
                    try FileManager.default.createDirectory(at: pendingDir, withIntermediateDirectories: true)
                }
                
                let downloadedFile = pendingDir.appendingPathComponent("bundle.tmp")
                if FileManager.default.fileExists(atPath: downloadedFile.path) {
                    try FileManager.default.removeItem(at: downloadedFile)
                }
                try FileManager.default.moveItem(at: location, to: downloadedFile)
                
                var finalZip = downloadedFile
                
                // Step A: Handle Differential Patch
                if let isPatch = updateJson["isPatch"] as? Bool, isPatch {
                    print("[\(self.TAG)] Applying differential patch...")
                    let currentBundleZip = otaDir.appendingPathComponent("bundle.zip")
                    let patchedZip = pendingDir.appendingPathComponent("bundle.patched.zip")
                    try self.applyPatch(oldFile: currentBundleZip, patchFile: downloadedFile, outputFile: patchedZip)
                    finalZip = patchedZip
                }
                
                // Step C: Verify Integrity
                let expectedHash = updateJson["hash"] as? String ?? ""
                if !expectedHash.isEmpty && !HashUtils.verifyHash(fileURL: finalZip, expectedHash: expectedHash) {
                    print("[\(self.TAG)] Hash verification failed!")
                    completion(false)
                    return
                }
                
                let signature = updateJson["signature"] as? String ?? ""
                if !signature.isEmpty, let publicKey = self.signingPublicKey {
                    if !SignatureVerifier.verifySignature(fileURL: finalZip, signatureBase64: signature, publicKeyHex: publicKey) {
                        print("[\(self.TAG)] Signature verification failed!")
                        completion(false)
                        return
                    }
                    print("[\(self.TAG)] Authenticity verified ✓")
                }
               
                // Step B: Handle Decryption
                if let isEncrypted = updateJson["isEncrypted"] as? Bool, isEncrypted {
                    print("[\(self.TAG)] Decrypting bundle...")
                    let decryptedZip = pendingDir.appendingPathComponent("bundle.decrypted.zip")
                    try self.decryptFile(at: finalZip, to: decryptedZip)
                    finalZip = decryptedZip
                }
                
                // Backup current
                if FileManager.default.fileExists(atPath: bundleDir.path) {
                    if FileManager.default.fileExists(atPath: previousDir.path) {
                        try FileManager.default.removeItem(at: previousDir)
                    }
                    try FileManager.default.moveItem(at: bundleDir, to: previousDir)
                }
                
                // Unzip or Move (Simplified)
                try FileManager.default.createDirectory(at: bundleDir, withIntermediateDirectories: true)
                // In reality, use a Zip library. For now, we assume the ZIP content is main.jsbundle
                // or we use a helper to extract.
                try self.unzip(file: finalZip, to: bundleDir)
                
                // Save metadata
                let metadata: [String: Any] = [
                    "version": updateJson["version"] as? String ?? "",
                    "hash": updateJson["hash"] as? String ?? "",
                    "release_id": updateJson["id"] as? String ?? updateJson["release_id"] as? String ?? "",
                    "installed_at": Date().timeIntervalSince1970
                ]
                if let metaData = try? JSONSerialization.data(withJSONObject: metadata) {
                    try? metaData.write(to: otaDir.appendingPathComponent("metadata.json"))
                }
                
                // Save zip for next diff
                let savedZip = otaDir.appendingPathComponent("bundle.zip")
                if FileManager.default.fileExists(atPath: savedZip.path) {
                    try FileManager.default.removeItem(at: savedZip)
                }
                try FileManager.default.copyItem(at: finalZip, to: savedZip)
                
                print("[\(self.TAG)] Update applied successfully.")
                
                // Report installation
                let releaseId = updateJson["id"] as? String ?? updateJson["release_id"] as? String ?? ""
                let isPatch = updateJson["isPatch"] as? Bool ?? false
                let downloadSize = try? FileManager.default.attributesOfItem(atPath: downloadedFile.path)[.size] as? Int64 ?? 0
                
                self.reportInstallation(releaseId: releaseId, status: "applied", isPatch: isPatch, downloadSize: downloadSize ?? 0)
                
                if let mandatory = updateJson["mandatory"] as? Bool, mandatory {
                    print("[\(self.TAG)] Mandatory update! Triggering reload...")
                    self.triggerReload()
                }
                
                completion(true)
                
            } catch {
                print("[\(self.TAG)] Update failed: \(error)")
                completion(false)
            }
        }.resume()
    }
    
    private func decryptFile(at source: URL, to destination: URL) throws {
        guard let keyHex = self.encryptionKey else {
            throw NSError(domain: "HotPatch", code: 400, userInfo: [NSLocalizedDescriptionKey: "Encryption key missing"])
        }
        
        let fileData = try Data(contentsOf: source)
        if fileData.count < 12 {
            throw NSError(domain: "HotPatch", code: 400, userInfo: [NSLocalizedDescriptionKey: "Invalid encrypted bundle"])
        }
        
        let nonceData = fileData.subdata(in: 0..<12)
        let ciphertext = fileData.subdata(in: 12..<fileData.count)
        
        let keyData = Data(hexString: keyHex)
        
        // CryptoKit Implementation
        #if canImport(CryptoKit)
        let key = SymmetricKey(data: keyData)
        let sealedBox = try AES.GCM.SealedBox(nonce: AES.GCM.Nonce(data: nonceData), ciphertext: ciphertext, tag: ciphertext.suffix(16))
        // Note: Our CLI appends tag to ciphertext if using standard Rust aes-gcm. 
        // We need to ensure the format matches. Standard AES-GCM tag is 16 bytes.
        // If the Rust implementation uses internal tag handling, ciphertext might include it at the end.
        
        // Adjusting ciphertext/tag split if necessary.
        let actualCiphertext = ciphertext.prefix(ciphertext.count - 16)
        let tag = ciphertext.suffix(16)
        let properBox = try AES.GCM.SealedBox(nonce: AES.GCM.Nonce(data: nonceData), ciphertext: actualCiphertext, tag: tag)
        
        let decryptedData = try AES.GCM.open(properBox, using: key)
        try decryptedData.write(to: destination)
        #else
        throw NSError(domain: "HotPatch", code: 500, userInfo: [NSLocalizedDescriptionKey: "CryptoKit not available"])
        #endif
    }
    
    private func reportInstallation(releaseId: String, status: String, isPatch: Bool, downloadSize: Int64) {
        guard let deviceId = UserDefaults.standard.string(forKey: "HotPatch_DeviceId") else { return }
        let urlString = "\(apiUrl)/installations"
        guard let url = URL(string: urlString) else { return }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue(appId, forHTTPHeaderField: "X-App-Key")
        
        let body: [String: Any] = [
            "device_id": deviceId,
            "release_id": releaseId,
            "status": status,
            "is_patch": isPatch,
            "download_size": downloadSize
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        // Use self.session (not URLSession.shared) so certificate pinning is applied
        session.dataTask(with: request).resume()
    }

    private func applyPatch(oldFile: URL, patchFile: URL, outputFile: URL) throws {
        print("[\(TAG)] Applying binary patch from \(oldFile.lastPathComponent)")
        try PatchUtils.applyPatch(oldURL: oldFile, patchURL: patchFile, newURL: outputFile)
    }
    
    private func unzip(file: URL, to directory: URL) throws {
        // Use the system `unzip` command available on all Apple platforms
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/unzip")
        process.arguments = ["-o", file.path, "-d", directory.path]
        process.standardOutput = FileHandle.nullDevice
        process.standardError = FileHandle.nullDevice
        
        try process.run()
        process.waitUntilExit()
        
        guard process.terminationStatus == 0 else {
            throw NSError(
                domain: "HotPatch",
                code: Int(process.terminationStatus),
                userInfo: [NSLocalizedDescriptionKey: "Failed to unzip bundle (exit code \(process.terminationStatus))"]
            )
        }
        
        print("[\(TAG)] Bundle extracted to \(directory.path)")
    }
    
    private func triggerReload() {
        DispatchQueue.main.async {
            // Post the standard React Native reload notification
            NotificationCenter.default.post(
                name: NSNotification.Name("RCTReloadNotification"),
                object: nil
            )
        }
    }

    // MARK: - URLSessionDelegate (Certificate Pinning)
    
    public func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        
        guard let serverTrust = challenge.protectionSpace.serverTrust,
              challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust else {
            completionHandler(.performDefaultHandling, nil)
            return
        }
        
        let host = challenge.protectionSpace.host
        
        // Check if we have pinning configured for this host (or if host ends with pinnedDomain)
        if let pinnedDomain = self.pinnedDomain,
           let expectedHash = self.pinnedPublicKeyHash,
           host.hasSuffix(pinnedDomain) {
            
            // Validate server trust
            var error: CFError?
            if SecTrustEvaluateWithError(serverTrust, &error) {
                if let serverCertificate = SecTrustGetCertificateAtIndex(serverTrust, 0) {
                    let serverPublicKey = SecCertificateCopyKey(serverCertificate)
                    if let publicKeyData = SecKeyCopyExternalRepresentation(serverPublicKey!, &error) as Data? {
                        let hash = SHA256.hash(data: publicKeyData)
                        let actualHash = hash.compactMap { String(format: "%02x", $0) }.joined()
                        
                        if actualHash.lowercased() == expectedHash.lowercased() {
                            print("[\(TAG)] Certificate pinning verified for \(host) ✓")
                            completionHandler(.useCredential, URLCredential(trust: serverTrust))
                            return
                        } else {
                            print("[\(TAG)] Certificate pinning FAILED for \(host). Expected: \(expectedHash), Got: \(actualHash)")
                            completionHandler(.cancelAuthenticationChallenge, nil)
                            return
                        }
                    }
                }
            }
            
            print("[\(TAG)] Certificate trust evaluation failed for \(host)")
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }
        
        // Default handling for other domains
        completionHandler(.performDefaultHandling, nil)
    }
}

extension Data {
    init(hexString: String) {
        var str = hexString
        if str.count % 2 != 0 { str = "0" + str }
        let bytes = stride(from: 0, to: str.count, by: 2).map {
            UInt8(str[str.index(str.startIndex, offsetBy: $0)...str.index(str.startIndex, offsetBy: $0 + 1)], radix: 16) ?? 0
        }
        self.init(bytes)
    }
}

extension String {
    subscript(bounds: CountableClosedRange<Int>) -> String {
        let start = index(startIndex, offsetBy: bounds.lowerBound)
        let end = index(startIndex, offsetBy: bounds.upperBound)
        return String(self[start...end])
    }
}
